### DISCUSSION TITLE
Date formatting control

### DISCUSSION THREAD

**Discussion by guykatz** [resteasy/resteasy#3591] [Q&A] [answered]
Hi all;
what is the portable way of controlling the default jaxrs date formatting (serialzation).
One of my jaxrs GET endpoints returns a 3rd party object (which I do not have control over). this object has has a java.util.Date property. 

I need to be able to change the default serialization for that date (control its json output formatting) because on the consumer of this endpoint expects a specific format.
thanks for your help!

**Comment by jamezp** [comment, marked as answer]
What version of RESTEasy are you using and which JSON provider?

**Reply by guykatz** [reply]
hi
I am running inside a wildfly 23 application server...
prefferably I want to stick to a jaxrs way (2.x) of doing things...
reasteasy version seems to be 3.15.1
not sure I ever chose a json provider so probably it uses the default?
thanks

**Reply by jamezp** [reply]
By default then, it should be using Jakarta JSON Processing. Are you referring to `java.util.Date` here? Or the new JSR 310 `javax.time.*`? I'm looking at https://jakarta.ee/specifications/jsonb/2.0/jakarta-jsonb-spec-2.0.html#dates to see how this could be controlled.

**Reply by jamezp** [reply]
Posted too soon. You can use a `ContextResolver` for the `Jsonb` instance.

```java
@Provider
public static class JsonbContextResolver implements ContextResolver<Jsonb> {
    @Override
    public Jsonb getContext(final Class<?> type) {
        return JsonbBuilder.create(new JsonbConfig().withDateFormat("dd-MMM-yyyy'T'HH:mm:SSS", Locale.getDefault()));
    }
}
```

Here is a Jakarta REST 3.1 example with the `SeBootstrap`.
```java
/*
 * JBoss, Home of Professional Open Source.
 *
 * Copyright 2023 Red Hat, Inc., and individual contributors
 * as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.wildfly.reproducer;

import java.time.ZonedDateTime;
import java.util.Date;
import java.util.Locale;
import java.util.Set;

import jakarta.enterprise.context.RequestScoped;
import jakarta.json.bind.Jsonb;
import jakarta.json.bind.JsonbBuilder;
import jakarta.json.bind.JsonbConfig;
import jakarta.ws.rs.ApplicationPath;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.SeBootstrap;
import jakarta.ws.rs.client.Client;
import jakarta.ws.rs.client.ClientBuilder;
import jakarta.ws.rs.core.Application;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import jakarta.ws.rs.ext.ContextResolver;
import jakarta.ws.rs.ext.Provider;

/**
 * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
 */
public class Main {

    public static void main(final String[] args) throws Throwable {
        System.setProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager");

        final SeBootstrap.Instance instance = SeBootstrap.start(RestActivator.class)
                .toCompletableFuture().get();

        try (Client client = ClientBuilder.newClient()) {
            final var target = client.target(instance.configuration().baseUriBuilder()).path("rest");
            try (
                    Response response = target.path("check").path("date").request().accept(MediaType.APPLICATION_JSON).
                            get()
            ) {
                System.out.println(response.getStatus());
                System.out.println(response.readEntity(String.class));
            }
            try (
                    Response response = target.path("check").path("date-time").request().accept(MediaType.APPLICATION_JSON).
                            get()
            ) {
                System.out.println(response.getStatus());
                System.out.println(response.readEntity(String.class));
            }
        }
        //Thread.currentThread().join();
        instance.stop().toCompletableFuture().get();
    }

    /**
     * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
     */
    @ApplicationPath("/rest")
    public static class RestActivator extends Application {
        @Override
        public Set<Class<?>> getClasses() {
            return Set.of(DateResource.class, JsonbContextResolver.class);
        }
    }

    /**
     * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
     */
    @Path("/check")
    @RequestScoped
    @Produces(MediaType.APPLICATION_JSON)
    public static class DateResource {

        @GET
        @Path("/date")
        public Date date() {
            return new Date();
        }

        @GET
        @Path("date-time")
        public ZonedDateTime zonedDateTime() {
            return ZonedDateTime.now();
        }
    }

    @Provider
    public static class JsonbContextResolver implements ContextResolver<Jsonb> {
        @Override
        public Jsonb getContext(final Class<?> type) {
            return JsonbBuilder.create(new JsonbConfig().withDateFormat("dd-MMM-yyyy'T'HH:mm:SSS", Locale.getDefault()));
        }
    }
}

```

**Reply by guykatz** [reply]
thanks for the ellaborate answer.
apperantly I still dont get how to get this one working
using my wildfly 23 app server, I added to my JAX RS Application extention class your static class (copy pasted form your example at the top)
that didnt work,
I then figured i probably need to register it and I am registering it via getSIngletons like so:
```
@ApplicationPath("/microstream")
public class RestApiApplicationMicrostream extends Application {
    Set<Object> singletons = Set.of(new MicrostreamRootResource(),new JsonbContextResolver());
    @Override
    public Set<Object> getSingletons() {
       return singletons;
    }
    @Provider
    public static class JsonbContextResolver implements ContextResolver<Jsonb> {
        @Override
        public Jsonb getContext(final Class<?> type) {
            return JsonbBuilder.create(new JsonbConfig().withDateFormat("dd-MMM-yyyy'T'HH:mm:SSS", Locale.getDefault()));
        }
    }
}
...
```
this also doesnt work.
the GET I am executing is:
```
    @GET 
    @Path("/raw/filesStatistics")
    @Produces("application/json")
    public ViewerStorageFileStatistics getRawStatistics() {
        return tps.restRawStatistics();
    }
```
the 3rd party object I am returning is: [https://docs.microstream.one/api/7.0/one/microstream/storage/restadapter/types/ViewerStorageFileStatistics.html](https://docs.microstream.one/api/7.0/one/microstream/storage/restadapter/types/ViewerStorageFileStatistics.html)
this object has a creationTime property which is what I strugle with - it is serialized as a long like so:
```
{
...
creationTime: 1683223199297,
...
}
```
I have compile about 30 different ways and couldnt achieve what I needed.
I must be missing something small or doing something wrong.
your help is appreciated.
thanks!

**Reply by jamezp** [reply]
Oh, you know I should try testing with Jakarta EE 8. There may have been some different defaults there. I'll have to setup a new example that runs on WildFly 23 to give it a try.

**Reply by guykatz** [reply]
Thanks.
I have seen some older posts say I should set a feature on the object mapper like so:
`objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)`
I will try to also see of this makes an impact

**Reply by jamezp** [reply]
That is used if you default to using Jackson instead of Jakarta JSON Binding.

**Reply by guykatz** [reply]
> That is used if you default to using Jackson instead of Jakarta JSON Binding.

Hmm....
Since I am not in full control pf the server maybe someone configured something.
Os there a programmatic way pf checking which binding is used?

**Comment by guykatz** [comment]
@jamezp hi
first of all, thanks for all the help here. I really appreciate it and I know it took some time...
I have found a reference in my server to jackson in code of one of the programmers so even though I dont see any special configurtation in the wilfdly config files to make jackson the provider I assumed it was actually the provider and was able to manipulate the format using the folloiwing provider code:
```
    public DateObjectMapperContextResolver() {
        mapper = new ObjectMapper();
        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS,false);
        mapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
    }
```
so all is good regarding this for now.
thanks again for your assistance.

**Reply by jamezp** [reply]
Awesome @guykatz! There are a few ways to trigger the server to use Jackson. One way is via the WildFly configuration. However, maybe you've got something annotated with a Jackson annotation.

Either way, that `DateObjectMapperContextResolver` looks correct to me. I'm glad you found it :)
